prov/usnic/src/usdf_ext.c

```diff
@@ -0,0 +1,159 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "fi.h"
+
+#include "usdf.h"
+#include "usnic_direct.h"
+#include "fi_ext_usnic.h"
+
+static int
+usdf_usnic_getinfo_v1(uint32_t version, struct fid_fabric *fabric,
+			struct fi_usnic_info *uip)
+{
+	struct usdf_fabric *fp;
+	struct usd_device_attrs *dap;
+
+	USDF_TRACE("\n");
+
+	fp = fab_ftou(fabric);
+	dap = fp->fab_dev_attrs;
+
+	/* this assignment was missing in libfabric v1.1.1 and earlier */
+	uip->ui_version = 1;
+
+	uip->ui.v1.ui_link_speed = dap->uda_bandwidth;
+	uip->ui.v1.ui_netmask_be = dap->uda_netmask_be;
+	strcpy(uip->ui.v1.ui_ifname, dap->uda_ifname);
+	uip->ui.v1.ui_num_vf = dap->uda_num_vf;
+	uip->ui.v1.ui_qp_per_vf = dap->uda_qp_per_vf;
+	uip->ui.v1.ui_cq_per_vf = dap->uda_cq_per_vf;
+
+	return 0;
+}
+
+static int usdf_usnic_getinfo_v2(uint32_t version, struct fid_fabric *ffabric,
+		struct fi_usnic_info *uip)
+{
+	struct usd_open_params params;
+	struct usd_device_attrs *dap;
+	struct usdf_fabric *fabric;
+	struct usd_device *dev;
+	int ret;
+	int i;
+
+	USDF_TRACE("\n");
+
+	fabric = fab_ftou(ffabric);
+	dap = fabric->fab_dev_attrs;
+
+	memset(&params, 0, sizeof(params));
+	params.flags = UOPF_SKIP_LINK_CHECK | UOPF_SKIP_PD_ALLOC;
+	params.cmd_fd = -1;
+	params.context = NULL;
+
+	ret = usd_open_with_params(dap->uda_devname, &params, &dev);
+	if (ret)
+		return -ret;
+
+	uip->ui_version = FI_EXT_USNIC_INFO_VERSION;
+
+	strcpy(uip->ui.v2.ui_devname, dap->uda_devname);
+	strcpy(uip->ui.v2.ui_ifname, dap->uda_ifname);
+	memcpy(uip->ui.v2.ui_mac_addr, dap->uda_mac_addr,
+			sizeof(dap->uda_mac_addr));
+
+	uip->ui.v2.ui_ipaddr_be = dap->uda_ipaddr_be;
+	uip->ui.v2.ui_netmask_be = dap->uda_netmask_be;
+	uip->ui.v2.ui_prefixlen = dap->uda_prefixlen;
+	uip->ui.v2.ui_mtu = dap->uda_mtu;
+	uip->ui.v2.ui_link_up = dap->uda_link_state;
+
+	uip->ui.v2.ui_device_id = dap->uda_device_id;
+	memcpy(uip->ui.v2.ui_firmware, dap->uda_firmware,
+			sizeof(dap->uda_firmware));
+
+	uip->ui.v2.ui_num_vf = dap->uda_num_vf;
+	uip->ui.v2.ui_cq_per_vf = dap->uda_cq_per_vf;
+	uip->ui.v2.ui_qp_per_vf = dap->uda_qp_per_vf;
+	uip->ui.v2.ui_intr_per_vf = dap->uda_intr_per_vf;
+	uip->ui.v2.ui_max_cq = dap->uda_max_cq;
+	uip->ui.v2.ui_max_qp = dap->uda_max_qp;
+
+	uip->ui.v2.ui_link_speed = dap->uda_bandwidth;
+	uip->ui.v2.ui_max_cqe = dap->uda_max_cqe;
+	uip->ui.v2.ui_max_send_credits = dap->uda_max_send_credits;
+	uip->ui.v2.ui_max_recv_credits = dap->uda_max_recv_credits;
+
+	for (i = 0; i < USD_CAP_MAX; i++) {
+		uip->ui.v2.ui_caps[i].capability = usd_capability(i);
+		uip->ui.v2.ui_caps[i].present = usd_get_cap(dev, i);
+	}
+
+	usd_close(dev);
+
+	return FI_SUCCESS;
+}
+
+static int usdf_usnic_getinfo(uint32_t version, struct fid_fabric *fabric,
+		struct fi_usnic_info *uip)
+{
```

Comment:
add `assert(FI_EXT_USNIC_INFO_VERSION == 2)` here to help our future selves not make a mistake